desc/equilibrium/equilibrium.py

```diff
@@ -843,60 +873,128 @@ def compute(
         if data is None:
             data = {}
 
-        # To avoid the issue of using the wrong grid for surface and volume averages,
-        # we first figure out what needed qtys are flux functions or volume integrals
-        # and compute those first on a full grid
         p = "desc.equilibrium.equilibrium.Equilibrium"
-        # If the user wants to compute x which depends on y which in turn depends on z,
-        # and they pass in y already computed in data, then we shouldn't need to compute
-        # z at all.
-        deps = list(
-            set(get_data_deps(names, obj=p, has_axis=grid.axis.size) + names)
-            - data.keys()  # subtract out y if already computed
+        deps = set(
+            get_data_deps(names, obj=p, has_axis=grid.axis.size, data=data) + names
         )
-        # TODO: replace this logic with `grid_type` from data_index
-        dep0d = [
-            dep
-            for dep in deps
-            if (data_index[p][dep]["coordinates"] == "") and (dep not in data)
-        ]
-        dep1dr = [
+
+        def need_src(name):
+            # Need to compute these on grid that is paired to the source grid, since
+            # the compute logic assume input data is evaluated on those coordinates.
+            # We exclude these from the depXdx sets below since the grids we will
+            # use to compute those dependencies are coordinate-blind.
+            # Example, "<L|r,a>" has coordinates="r", but requires computing on
+            # field line following source grid.
+            return bool(data_index[p][name]["source_grid_requirement"])
+
+        # Need to call _grow_seeds so that some other quantity like K = 2 * <L|r,a>,
+        # which does not need a source grid to evaluate, does not compute <L|r,a> on a
+        # grid that does not follow field lines.
+        need_src_deps = _grow_seeds(p, set(filter(need_src, deps)), deps)
```

Comment:
It's the other way.

> Is there somewhere in this function call that ensures that these dependencies themselves do not require source_grid

That's precisely `need_src_deps`.

`need_src_deps` is
1. anything that needs to get computed that also
2. has a dependency which requires a special non-DESC-coordinate grid to compute

We inform the code of 2. by registering the compute function with a `source_grid` `kwarg ` and include whatever requirements are expected from the `source_grid` in that `kwarg`.

The effective ripple needs to do integration along field lines, so it requires a source grid in rho, alpha, zeta coordinates. This is so that we know which nodes in the DESC grid are along a field line (fixed rho, alpha). We also typically need to access the original source grid so that we know how far apart the nodes along a field line are.

---
Suppose we want to compute the volume average of the effective ripple. The compute function
```python
data=["effective ripple"],
def _compute_volume_average_ripple(....):
    data["volume average ripple"] = jnp.sum(jacorbian weights * data["effective ripple"])
    return data
```
does not directly need any special grid, however the dependency "effective ripple", certainly does. We need to ensure that we try to compute `volume average ripple` on a QuadratureGrid, because that would cause the dependency "effective ripple" to be computed on QuadratureGrid as well, giving wrong results. So we collect all such quantities that should be computed on the special grid in `need_src_deps`.

(Whether or not volume average ripple is important to anyone is irrelevant, since there are other examples where this issue arises, so it needs to be resolved). There is also no way to resolve this in a recursive manner that is smart about swapping grids and broadcasting the dependencies etc. I have a local branch that attempted this, and it reduces to this one since in general one needs to walk up the entire dependency tree with a quadrature grid first to get all the quantities that are constant throuhout the plasma volume to compute the special grid quantities.